File: games_js/WheelOfFortune/wheelOfFortune.js
let categories = {
	'Fun & Games':
		'Competitive Shuffleboard|Neighborhood Get-Together|Skateboards & Rollerblades|Snowshoeing & Snowboarding|Ice-Carving Competition|Snowbiking & Airboarding|City-Center Horseraces|Hamletscenen Festival|Lion-Dance Competition|Snowboarding & Sledding|Traditional Macaroons|Building Sandcastles|Festive Celebrations|Fingerprints Drawing|RC Airplane Racing|Alpine Snowboarding|Contestant Searches|Festive Celebration|Freestyle Wrestling|Interactive Puzzles|Mini-Golf Tournament|Playing Racquetball|Back-Road Bicycling|Complicated Puzzle|Fast-Pitch Softball|Freshwater Fishing|Outdoor Recreation|Playing Backgammon|Playing Pictionary|Water Balloon Fight|Weeklong Festivals|Chess Competition|Crossword Puzzles|Equestrian Sports|Football & Baseball|Going Parasailing|Halloween Hayride|Indoor Volleyball|Juggling Beanbags|Medieval Festival|Playing Asteroids|Playing Hopscotch|Playing Solitaire|Pothole Exploring|Softball & Baseball|Two-Story Carousel|Wheelbarrow Races|Adventure Racing|Ballroom Dancing|Beach Volleyball|Childhood Heroes|Christmas Crafts|Demolition Derby|Doing Handstands|Dungeons & Dragons|Fraternity Prank|Fun Brainteasers|Gorgeous Fishing|Habanos Festival|Hula Competition|Indoor Go-Carting|Inflatable Slide|Japanese Archery|Knock-Knock Jokes|Narrated Cruises|Ping-Pong Paddles|Playing Dominoes|Playing Jeopardy!|Playing Lacrosse|Playing Peekaboo|Playing Scrabble|Potato-Sack Races|Riding Piggyback|Spitting Contest|Street Carnivals|Twenty Questions|Urban Spelunking|Wheelbarrow Race|Amazing History|Anderlecht Fair|Big-Wave Surfing|Boggle & Scrabble|Burping Contest|Community Chest|Country Dancing|Croquet Mallets|Downhill Skiing|Dragon-Boat Race|Family Cookouts|Gaelic Football'.split(
			'|'
		)
};

let board = [];
let score = 0;
let emptyBoxes, phrase, words, buzzed, phrases;

let bigBuzzer = `
|⎺|__  _   _ ___________ _ __
| '_ \\| | | |_  /_  / _ \\ '__|
| |_) | |_| |/ / / /  __/ |
|_.__/ \\__,_/___/___\\___|_|`.slice(1);

function displayScore() {
	text('Score: ' + score, 17, 14);
}

async function pickNewPhrase() {
	await erase();

	displayScore();

	let category = selectedCategory;

	if (selectedCategory == 'Shuffle!') {
		let cats = Object.keys(categories);
		category = cats[Math.floor(Math.random() * cats.length)];
	}

	text('The category is ' + category, 2, 2);

	let phrases = categories[category];
	phrase = phrases[Math.floor(Math.random() * phrases.length)];
	words = phrase.split(' ');
	log(words);

	board = [];

	emptyBoxes = 0;
	/* Display all the boxes */
	for (let i = 0; i < words.length; i++) {
		let word = words[i];
		board.push([]);
		for (let j = 0; j < word.length; j++) {
			let character = word[j];
			if (character != '-' && character != '&') {
				board[i].push(' ');
				emptyBoxes++;
			} else {
				board[i].push(character);
				text(character, 5 + i * 3, 3 + j * 3);
			}
			textRect(4 + i * 3, 2 + j * 3, 3, 3);
		}
	}

	/* Create the buzzer button */
	buzzed = false;
	button(bigBuzzer, 18, 5, buzz);
	addLetter();
}

async function buzz() {
	buzzed = true;
	let guess = await prompt('Guess the phrase:', 18);

	if (typeof guess == 'string' && guess.toLowerCase() == phrase.toLowerCase()) {
		score += emptyBoxes;
		displayScore();
		await alert('Correct!', 18);
		pickNewPhrase();
	} else {
		score -= 3;
		displayScore();
		await alert('Wrong! Try again.', 18);
		buzzed = false;
		button(bigBuzzer, 18, 5, buzz);
		addLetter();
	}
}

async function gameOver() {
	score -= 3;
	displayScore();
	await alert('Out of time! The phrase was "' + phrase + '"', 17);
	pickNewPhrase();
}

/* Add a letter to a random empty box */
async function addLetter() {
	if (emptyBoxes <= 0) {
		gameOver();
		return;
	}

	if (buzzed) return;

	let rand = Math.ceil(Math.random() * emptyBoxes);
	let emptyBoxCount = 0;

	for (let i = 0; i < words.length; i++) {
		let word = words[i];
		for (let j = 0; j < word.length; j++) {
			if (board[i][j] == ' ') {
				emptyBoxCount++;
				if (emptyBoxCount != rand) continue;

				board[i][j] = words[i][j];
				text(words[i][j], 5 + i * 3, 3 + j * 3);
				emptyBoxes--;
				await delay(1250);
				addLetter();
				return;
			}
		}
	}
}

async function selectScreen() {
	await erase();

	let description = 'Guess the phrase as quick as you can! Select a Category:';
	await text(description, 2, 2, 36);

	let cats = Object.keys(categories);
	cats.push('Shuffle!');
	for (let i = 0; i < cats.length; i++) {
		let categoryName = cats[i];
		button(categoryName.slice(0, 17).padEnd(17), 5 + (i % 17), i < 17 ? 2 : 21, async () => {
			await erase();
			selectedCategory = categoryName;
			await alert('You selected "' + selectedCategory + '".\n\nGood luck!');
			pickNewPhrase();
		});
	}
}

async function setup() {
	// source: https://gist.github.com/michaelmotzkus/de82e06c8538399909103108049788b9
	Object.assign(categories, await (await fetch(QuintOS.dir + '/categories.json')).json());
	selectScreen();
}

File: games_java/WheelOfFortune/WheelOfFortune.java
package games_java.WheelOfFortune;

import static games_java.QuintOS.*;

import java.io.File;
import java.io.FileNotFoundException;
import java.util.ArrayList;
import java.util.Scanner;

public class WheelOfFortune {
	// the category is "Fun and Games"
	String bigBuzzer = """
			|⎺|__  _   _ ___________ _ __
			| '_ \\| | | |_  /_  / _ \\ '__|
			| |_) | |_| |/ / / /  __/ |
			|_.__/ \\__,_/___/___\\___|_|""".substring(1);

	String[] phrases;
	String phrase;
	String[] words;
	char[][] board;
	boolean buzzed;
	Scanner sc;
	int score;

	public WheelOfFortune() {
		sc = new Scanner(System.in);
		score = 0;
		setup();
	}

	void setup() {
		String dir = System.getProperty("user.dir") + "/games_java/WheelOfFortune";
		File phrasesFile = new File(dir + "/phrases.txt");
		delay(2000);
		Scanner fileScanner;
		try {
			fileScanner = new Scanner(phrasesFile);
		} catch (FileNotFoundException e) {
			e.printStackTrace();
			return;
		}
		ArrayList<String> phrasesList = new ArrayList<String>();
		while (fileScanner.hasNextLine()) {
			phrasesList.add(fileScanner.nextLine());
		}
		fileScanner.close();
		/*
		 * Make an array of phrases, pick a random phrase, and split pharse into an
		 * array of words
		 */
		phrases = new String[phrasesList.size()];
		for (int i = 0; i < phrasesList.size(); i++) {
			phrases[i] = phrasesList.get(i);
		}
		phrase = phrases[(int) (Math.random() * phrases.length)];
		words = phrase.split(" ");
		log(words);

		/* Make a board array to represent the letters in the phrase */
		// words -> ['Community', 'Chest']
		// board -> [
		// [' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' '],
		// [' ', ' ', ' ', ' ', ' ']
		// ]
		board = new char[words.length][];

		for (int wordCounter = 0; wordCounter < words.length; wordCounter++) {
			board[wordCounter] = new char[words[wordCounter].length()];
			for (int letterCounter = 0; letterCounter < words[wordCounter].length(); letterCounter++) {
				board[wordCounter][letterCounter] = ' ';
			}
		}
		log(board);

		buzzed = false;

		/* Create the buzzer button */
		button(bigBuzzer, 18, 5, () -> {
			this.buzz();
		});

		text("Score: " + score, 17, 10);
		displayBoxes();
		// addLetter();
	}

	/* Display all the boxes for the phrase */
	void displayBoxes() {
		String boxes = "";
		for (int wordCounter = 0; wordCounter < words.length; wordCounter++) {
			String word = words[wordCounter];
			boxes += "┌─┐".repeat(word.length()) + "\n";
			for (int letterCounter = 0; letterCounter < word.length(); letterCounter++) {
				boxes += "│" + board[wordCounter][letterCounter] + "│";
			}
			boxes += "\n" + "└─┘".repeat(word.length()) + "\n";
		}
		System.out.println(boxes);
	}

	void buzz() {
		buzzed = true;
		System.out.println("What's your guess?");
		String guess = sc.nextLine();
		if (guess != null && guess.equals(phrase)) {
			int lettersLeft = 0;
			for (int w = 0; w < words.length; w++) {
				for (int l = 0; l < words[w].length(); l++) {
					if (board[w][l] == ' ') {
						lettersLeft++;
					}
				}
			}
			score += lettersLeft;
			erase();
			setup();
			return;
		} else {
			score--;
		}
		text("Score: " + score, 17, 10);
		button(bigBuzzer, 18, 5, () -> {
			this.buzz();
		});
		buzzed = false;
		addLetter();
	}

	/* Add a letter to a random empty box */
	void addLetter() {
		ArrayList<int[]> avail = new ArrayList<int[]>();

		for (int w = 0; w < words.length; w++) {
			for (int l = 0; l < words[w].length(); l++) {
				if (board[w][l] == ' ') {
					avail.add(new int[] { w, l });
				}
			}
		}
		log(avail);

		int[] coord = avail.get((int) (Math.random() * avail.size()));

		int whichWord = coord[0];
		int whichLetter = coord[1];

		board[whichWord][whichLetter] = words[whichWord].charAt(whichLetter);
		displayBoxes();

		try {
			Thread.sleep(1000);
		} catch (InterruptedException e) {
			// System.err.println("delay failed");
		}

		// if (avail.size() > 1 && !buzzed) {
		// addLetter();
		// }
		// if (avail.size() == 1) {
		// score -= 3;
		// erase();
		// setup();
		// }
	}

	public static void main(String[] args) {
		new WheelOfFortune();
	}

}
